### PR TITLE
Fixed JNI codes to make it compilable with VisualStudio and runnable …

### PR DESCRIPTION
**Pull request by hideki**
…on Windows platform

- VC++ does not allow dynamic array, changed to use new/delete for VC++ https://msdn.microsoft.com/en-us/library/eff825eh.aspx
-  VC++ does not all (type){...}
- __func__ -> __FUNCTION__ advised by @jim

### CODE REVIEW

**Comment by hideki** [comment]
@borrrden I assigned this PR to @snej. Can you also review the PR? Thanks!

File: Java/jni/native_document.cc [hunk 1]
```diff
@@ -262,7 +262,11 @@ JNIEXPORT jint JNICALL Java_com_couchbase_cbforest_Document_insertRevisionWithHi
     {
         // Convert jhistory, a Java String[], to a C array of C4Slice:
         jsize n = env->GetArrayLength(jhistory);
+#ifdef _MSC_VER
+        C4Slice* history = new C4Slice[n];
```
Comment (snej) [inline Java/jni/native_document.cc]: If we're going to use dynamic allocation, this should be a vector -- `std::vector<C4Slice> history(n)`.

File: Java/jni/native_glue.cc [hunk 2]
```diff
@@ -104,11 +104,21 @@ namespace forestdb {
         jstring toJString(JNIEnv *env, C4Slice s) {
             if (s.buf == NULL)
                 return NULL;
+#ifdef _MSC_VER
+            char* utf8Buf = new char[s.size + 1];
```
Comment (snej) [inline Java/jni/native_glue.cc]: Use a `std::string` instead:

```c++
std::string utf8Buf((char*)s.buf, s.size);
jstring ret = env->NewStringUTF(utf8Buf.c_str());
```

File: Java/jni/native_glue.hh [hunk 3]
```diff
@@ -34,7 +34,11 @@ public:
     { s._slice = slice::null; }
 
     operator slice()    {return _slice;}
+#ifdef _MSC_VER
+    operator C4Slice()  {return {_slice.buf, _slice.size};}
```
Comment (snej) [inline Java/jni/native_glue.hh]: You don't need to conditionalize this. Just change the existing implementation to use the standard syntax.

File: Java/jni/native_glue.hh [hunk 4]
```diff
@@ -55,7 +59,11 @@ public:
...
```
Comment (snej) [inline Java/jni/native_glue.hh]: Same as above.

File: Java/jni/native_view.cc [hunk 5]
```diff
@@ -147,8 +147,13 @@ JNIEXPORT void JNICALL Java_com_couchbase_cbforest_View_emit(JNIEnv *env, jobjec
     C4Document* doc = (C4Document*)documentHandler;
     size_t count = env->GetArrayLength(jkeys);
     jlong *keys   = env->GetLongArrayElements(jkeys, NULL);
+#ifdef _MSC_VER
+    C4Key** c4keys = new C4Key*[count];
```
Comment (snej) [inline Java/jni/native_view.cc]: Please use vectors.

File: Java/jni/native_view.cc [hunk 6]
```diff
@@ -239,14 +257,22 @@ JNIEXPORT jlong JNICALL Java_com_couchbase_cbforest_View_query__JJJZZZ_3J
         (bool)inclusiveEnd,
         NULL,
         NULL,
+#ifdef _MSC_VER
```
Comment (snej) [inline Java/jni/native_view.cc]: No need to conditionalize this, just change it.

**Comment by borrrden** [comment]
I agree with Jens' comments.

**Comment by hideki** [comment]
@snej Could you please review PR again? Thanks!

**Comment by hideki** [comment]
@snej, Sorry I forgot to fix one more place. Please forget previous my comment.

**Comment by hideki** [comment]
@snej Could you please review PR again? Thanks a lot!

**Pull request merged**